service/turn.proxy/src/main/java/org/kaazing/gateway/service/turn/proxy/TurnProxyAcceptHandler.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright 2007-2016, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.gateway.service.turn.proxy;
+
+import org.apache.mina.core.future.ConnectFuture;
+import org.apache.mina.core.future.IoFutureListener;
+import org.apache.mina.core.session.IoSession;
+import org.apache.mina.core.session.IoSessionInitializer;
+import org.kaazing.gateway.security.SecurityContext;
+import org.kaazing.gateway.service.ServiceContext;
+import org.kaazing.gateway.service.ServiceProperties;
+import org.kaazing.gateway.service.proxy.AbstractProxyAcceptHandler;
+import org.kaazing.gateway.service.proxy.AbstractProxyHandler;
+import org.kaazing.gateway.service.turn.proxy.stun.StunCodecFilter;
+import org.kaazing.gateway.service.turn.proxy.stun.StunProxyMessage;
+import org.kaazing.gateway.service.turn.proxy.stun.attributes.*;
+import org.kaazing.gateway.util.turn.TurnUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.security.Key;
+import java.security.KeyStoreException;
+import java.util.Base64;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TurnProxyAcceptHandler extends AbstractProxyAcceptHandler {
+
+    static final Logger LOGGER = LoggerFactory.getLogger(TurnProxyAcceptHandler.class);
+    public static final String TURN_STATE_KEY = "turn-state";
+
+    // Configuration properties
+    public static final String PROPERTY_MAPPED_ADDRESS = "mapped.address";
+    public static final String PROPERTY_KEY_ALIAS = "key.alias";
+    public static final String PROPERTY_KEY_PASS_FILE = "key.password-file";
+    public static final String PROPERTY_KEY_ALGORITHM = "key.algorithm";
+
+    private String connectURI;
+    private Key sharedSecret;
+    private String keyAlgorithm;
+
+    private ConcurrentHashMap<String, String> currentTransactions = new ConcurrentHashMap<>();
```

Comment:
I think this is a bug... 

1) I think hypothetically two stun sessions might create the same transaction ID at the same time.  One would override the other.

2) Concurrent data structures have a large overhead in terms of performance, especially when done per every message multiple times through decoders and encoders.   Perhaps, a better approach would to create the transaction map per a session and store it on the session object (via Attribute?) and then accesses via the session.  In such a case, access to it should be thread safe (if thread alignment is working) and thus it should all be good.    Note, the RFC says 

> 
   At any time, a client MAY have multiple outstanding STUN requests
   with the same STUN server (that is, multiple transactions in
   progress, with different transaction IDs).  Absent other limits to
   the rate of new transactions (such as those specified by ICE for
   connectivity checks or when STUN is run over TCP), a client SHOULD
   space new transactions to a server by RTO and SHOULD limit itself to
   ten outstanding transactions to the same server.

so perhaps a good initial capacity for the hash map may be 10...?